api/server/router/build/build_routes.go

```diff
@@ -159,22 +154,31 @@ func (br *buildRouter) postBuild(ctx context.Context, w http.ResponseWriter, r *
 
 	buildOptions.AuthConfigs = authConfigs
 
-	out = output
+	var out io.Writer = output
 	if buildOptions.SuppressOutput {
 		out = notVerboseBuffer
 	}
 	stdout := &streamformatter.StdoutFormatter{Writer: out, StreamFormatter: sf}
 	stderr := &streamformatter.StderrFormatter{Writer: out, StreamFormatter: sf}
 
-	closeNotifier := make(<-chan bool)
+	finished := make(chan struct{})
+	defer close(finished)
 	if notifier, ok := w.(http.CloseNotifier); ok {
-		closeNotifier = notifier.CloseNotify()
+		notifyContext, cancel := context.WithCancel(ctx)
+		closeNotifier := notifier.CloseNotify()
+		go func() {
+			select {
+			case <-closeNotifier:
+				cancel()
+			case <-finished:
+			}
+		}()
+		ctx = notifyContext
```

Comment:
Maybe something for a different PR, but it seems like this belongs in `makeHTTPHandler` so it works for all routes. This would benefit pull/push as well. Right now they cancel their context in daemon-level code if a write to the HTTP response fails. It would be cleaner to use `CloseNotifier`, but it shouldn't have to be duplicated for each route.